crates/ruma-server-util/src/authorization.rs

```diff
@@ -0,0 +1,279 @@
+//! Common types for implementing federation authorization.
+
+use headers::{authorization::Credentials, HeaderValue};
+
+use ruma_common::{OwnedServerName, OwnedServerSigningKeyId};
+use tracing::debug;
+use yap::{IntoTokens, TokenLocation, Tokens};
+
+/// Typed representation of an `Authorization` header of scheme `X-Matrix`, as defined in the
+/// [Matrix Server-Server API][spec]. Includes an implementation of
+/// [`headers::authorization::Credentials`] for automatically handling the encoding and decoding
+/// when using a web framework that supports typed headers.
+///
+/// [spec]: https://spec.matrix.org/v1.3/server-server-api/#request-authentication
+#[non_exhaustive]
+pub struct XMatrix {
+    /// The server name of the sending server.
+    pub origin: OwnedServerName,
+    /// The server name of the receiving sender. For compatibility with older servers, recipients
+    /// should accept requests without this parameter, but MUST always send it. If this property is
+    /// included, but the value does not match the receiving server's name, the receiving server
+    /// must deny the request with an HTTP status code 401 Unauthorized.
+    pub destination: Option<OwnedServerName>,
+    /// The ID - including the algorithm name - of the sending server's key that was used to sign
+    /// the request.
+    pub key: OwnedServerSigningKeyId,
+    /// The signature of the JSON.
+    pub sig: String,
+}
+
+impl XMatrix {
+    /// Construct a new X-Matrix Authorization header.
+    pub fn new(
+        origin: OwnedServerName,
+        destination: Option<OwnedServerName>,
+        key: OwnedServerSigningKeyId,
+        sig: String,
+    ) -> Self {
+        Self { origin, destination, key, sig }
+    }
+}
+
+fn parse_token<'a>(tokens: &mut impl Tokens<Item = &'a u8>) -> Option<Vec<u8>> {
+    tokens.optional(|t| {
+        let token: Vec<u8> = t.tokens_while(|c| is_tchar(**c)).copied().collect();
+        if !token.is_empty() {
+            Some(token)
+        } else {
+            debug!("Returning early because of empty token at {}", t.location().offset());
+            None
+        }
+    })
+}
+
+fn parse_token_with_colons<'a>(tokens: &mut impl Tokens<Item = &'a u8>) -> Option<Vec<u8>> {
+    tokens.optional(|t| {
+        let token: Vec<u8> = t.tokens_while(|c| is_tchar(**c) || **c == b':').copied().collect();
+        if !token.is_empty() {
+            Some(token)
+        } else {
+            debug!("Returning early because of empty token at {}", t.location().offset());
+            None
+        }
+    })
+}
+
+fn parse_quoted<'a>(tokens: &mut impl Tokens<Item = &'a u8>) -> Option<Vec<u8>> {
+    tokens.optional(|t| {
+        t.token(&b'"').then(|| ())?;
```

Comment:
I'm not a fan of this kind of `bool::then` usage, it's a bit too opaque. I'd prefer

```suggestion
        if !t.token(&b'"') {
            return None;
        }
```